### PR TITLE
Better gamepad support

### PR DESCRIPTION
**Pull request by rumly111**
#### Summary
Interface "Better gamepad support"

#### Purpose of change
Gamepad sticks/triggers are handled properly now. Diagonal movement with left stick works properly. Also long-pressing gamepad button/stick/trigger generates repeated input events (with some exceptions).

#### Describe the solution
1. Switched from Joystick API to GameController API
2. Moved all gamepad handling to separate file
3. Made a SDL timer that generates GAMEPAD_SCHEDULER events every 50ms, it's required to continuously generate gamepad button events for long-pressed buttons.
4. Increased the number of buttons to 30
5. Left stick sends JOY_LEFT,JOY_RIGHT,...,JOY_LEFTUP,JOY_LEFTDOWN button codes. Right stick sends JOY_21..JOY_28. LTrigger = JOY_29, RTrigger = JOY_30.
6. To make menuing easier, some gamepad buttons are hardcoded and can't be configured by user:
    6.1 DPAD generates keyboard ARROW key events
    6.2 BACK generates Escape
    6.3 START generates Enter
7. Added some gamepad default bindings.

#### Describe alternatives you've considered
The player could disable joystick support and use joy2key or similar tool. For example this script made by me https://pastebin.com/94G0Fxi3

#### Testing
I only tested it with my xbox360 controller, and it's playable. I'd like it if somebody with a PS dualshock could test it. I'm also not certain how the game will behave if the joystick doesn't support the gamecontroller specification.

#### Additional context
This solution is not perfect. The main problem with the gamepad is that there is too few buttons. I tried making defaults so you can walk around and fight with the gamepad. As for crafting/building/etc, you'd still need to switch to keyboard.

### CODE REVIEW

File: src/gamepad.cpp [hunk 1]
```diff
@@ -0,0 +1,374 @@
+#include "gamepad.h"
```
Comment (ZhilkinSerg) [inline src/gamepad.cpp:1]: I think it would make sense to rename the file (and its corresponding header file) to `sdl_gamepad`.

Comment (rumly111) [reply src/gamepad.cpp:1]: not to me

**Comment by rumly111** [comment]
I don't get it, why does compilation fail?
```
src/sdl_wrappers.h:13:13: fatal error: SDL.h: No such file or directory
```
I tried replacing `#include <SDL.h>` with `#include "sdl_wrapper.h"`, but it still can't find SDL headers :-/

File: src/gamepad.cpp [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+
+static Uint32 timer_func( Uint32 interval, __attribute__( ( unused ) ) void *param )
+{
+    SDL_Event event;
+    SDL_UserEvent userevent;
+
+    userevent.type = GAMEPAD_SCHEDULER;
+    userevent.code = 0;
+    userevent.data1 = NULL;
+    userevent.data2 = NULL;
```
Comment (ZhilkinSerg) [inline src/gamepad.cpp:72]:
```suggestion
    userevent.data1 = nullptr;
    userevent.data2 = nullptr;
```

File: src/gamepad.cpp [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+
+struct task_t {
+    Uint32 when;
+    int button;
+    int counter;
+    int state;
+};
+
+constexpr int max_tasks           = max_buttons + max_sticks + max_triggers + 1;
+constexpr int buttons_task_index  = 0;
```
Comment (ZhilkinSerg) [inline src/gamepad.cpp:47]:
```suggestion
// constexpr int buttons_task_index  = 0;
```

File: src/gamepad.cpp [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+#define dbg(x) DebugLog((x),D_SDL) << __FILE__ << ":" << __LINE__ << ": "
+
+extern input_event last_input;
+
+namespace gamepad
+{
+
+constexpr int max_triggers = 2;
+constexpr int max_sticks = 2;
+constexpr int max_axis = 6;
```
Comment (ZhilkinSerg) [inline src/gamepad.cpp:14]:
```suggestion
// constexpr int max_axis = 6;
```

File: src/gamepad.cpp (hunk 1 above)
Comment (ZhilkinSerg) [inline src/gamepad.cpp:1]:
```suggestion
#if defined(TILES)
#include "gamepad.h"
```

File: src/gamepad.cpp [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+    Uint32 now = event.user.timestamp;
+    for( auto &task : all_tasks )
+        if( task.counter && task.when <= now ) {
+            send_input( task.button );
+            task.counter += 1;
+            task.when = now + repeat_interval;
+        }
+}
+
+} // namespace gamepad
```
Comment (ZhilkinSerg) [inline src/gamepad.cpp:374]:
```suggestion
} // namespace gamepad

#endif // TILES
```

File: src/gamepad.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+#ifndef GAMEPAD_H
+#define GAMEPAD_H
+
```
Comment (ZhilkinSerg) [inline src/gamepad.h:4]:
```suggestion

#if defined(TILES)
```

File: src/gamepad.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+void init();
+void quit();
+void handle_axis_event( SDL_Event &event );
+void handle_hat_event( SDL_Event &event );
+void handle_button_event( SDL_Event &event );
+void handle_scheduler_event( SDL_Event &event );
+SDL_GameController *get_controller();
+
+}
+
```
Comment (ZhilkinSerg) [inline src/gamepad.h:22]:
```suggestion

#endif // TILES
```

**Comment by ZhilkinSerg** [comment]
> I don't get it, why does compilation fail?
> 
> ```
> src/sdl_wrappers.h:13:13: fatal error: SDL.h: No such file or directory
> ```
> 
> I tried replacing `#include <SDL.h>` with `#include "sdl_wrapper.h"` in gamepad.h, but it still can't find SDL headers :-/ Do I need to somehow register the newly added gamepad.cpp/.h ? It compiles on my laptop without any additional steps.

You need to wrap any SDL-related code into 

```
#if defined(TILES)
...
#endif // TILES
```

File: src/gamepad.cpp (hunk 3 above)
Comment (rumly111) [reply src/gamepad.cpp:47]: even if the variable `button_task_index` is not used directly, it's more like a comment (or documentation) to remind me where those button tasks start in the whole bunch.

**Comment by rumly111** [comment]
> > I don't get it, why does compilation fail?
> > ```
> > src/sdl_wrappers.h:13:13: fatal error: SDL.h: No such file or directory
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I tried replacing `#include <SDL.h>` with `#include "sdl_wrapper.h"` in gamepad.h, but it still can't find SDL headers :-/ Do I need to somehow register the newly added gamepad.cpp/.h ? It compiles on my laptop without any additional steps.
> 
> You need to wrap any SDL-related code into
> 
> ```
> #if defined(TILES)
> ...
> #endif // TILES
> ```

Oh, the ncurses version. I knew it's gonna bite me :) I'm gonna add `#if defined(TILES)` . As for those unused variables and functions, they can be treated as comments. And the unused functions can be used somewhere in the future, it's a pity to delete them.

**Comment by BrettDong** [comment]
> As for those unused variables and functions, they can be treated as comments. And the unused functions can be used somewhere in the future, it's a pity to delete them.

Warning-free is mandatory for code merge.

**Comment by rumly111** [comment]
some builds failed, but they don't seem related to my changes

**Comment by rumly111** [comment]
thanks @ZhilkinSerg

**Comment by liminalRoar** [comment]
I just saw this pr. what do you think of #56996 ? Not sure if you play CoQ or have tried gamepad on there. I think some of these ideas would make gamepad gameplay even better though!

**Pull request merged**